issue 3109144: Add Unix Domain Socket support in SOCK_STREAM and SOCK_DGRAM

EldarShalev opened the pull request:
Adding Unix Domain Socket support in SOCK_STREAM and SOCK_DGRAM
Signed-off-by: Eldar Shalev <eldarsh@nvidia.com>
Author:    Eldar Shalev <eldarsh@nvidia.com>

Adding new support for Unix Domain Socket on 2 modes - SOCK_STREAM and SOCK_DGRAM.
Binding automatically to the given path with suffix  "_client" or override using "--client_addr" as a parameter.
Example:
server side:
sockperf sr --addr PATH_TO_UDS_NAME
client side:
option 1: sockperf pp --addr PATH_TO_UDS_NAME
option 2: sockperf pp --addr PATH_TO_UDS_NAME --client_addr BINDING_NAME

Notes:
* Defualt is SOCK_DGRAM, for SOCK_STREAM use '--tcp'
* '--addr' was added in addition to '--ip'
* '--client_addr' was added in addition to '--client_ip'

Commit:
issue 3109144: Adding Unix Domain Socket support in SOCK_STREAM and SOCK_DGRAM

Signed-off-by: Eldar Shalev <eldarsh@nvidia.com>
Author:    Eldar Shalev <eldarsh@nvidia.com>

Adding new support for Unix Domain Socket on 2 modes - SOCK_STREAM and SOCK_DGRAM.
Binding automatically to the given path with suffix  "_client" or override using "--client_addr" as a parameter.
Example:
server side:
sockperf sr --addr PATH_TO_UDS_NAME
client side:
option 1: sockperf pp --addr PATH_TO_UDS_NAME
option 2: sockperf pp --addr PATH_TO_UDS_NAME --client_addr BINDING_NAME

Notes:
* Defualt is SOCK_DGRAM, for SOCK_STREAM use '--tcp/--stream'
* '--addr' was added in addition to '--ip'
* '--stream' was added in addition to '--tcp'
* '--client_addr' was added in addition to '--client_ip'

swx-jenkins3 commented:
Can one of the admins verify this patch?

agalanin-at-nvidia commented on src/defs.h:
```diff
@@ -482,6 +483,7 @@ struct sockaddr_store_t {
         sa_family_t ss_family;
         sockaddr_in addr4;
         sockaddr_in6 addr6;
+        sockaddr_un addr_un;
```
`sockaddr_un` is longer than 100 bytes so this structure size is 10x to previous revision. This may result in performance degradations in socket address handling code (e.g. statistics).

EldarShalev replied:
@agalanin-at-nvidia should we pass this struct in another way? 
@rosenbaumalex  any ideas how to proceed?

agalanin-at-nvidia replied:
You need to carefully check all the places when address (sockaddr_store_t or IPAddress) is copied or used as a key in containers. At least you need to convert IPAddress to something memory-efficient than does not copy addr_un for IP addressed.

EldarShalev replied:
@agalanin-at-nvidia  I guess it's important only in data path?  
if so, I can't find places where it's copy. 
AFAIK in both msg_sendto / receive_pending_data functions there is no copy of sockaddr_store_t.

agalanin-at-nvidia replied:
Some places I found:
* https://github.com/Mellanox/sockperf/blob/f80980feaf5334532d21df621c5791202f19dfab/src/server.h#L285
* https://github.com/Mellanox/sockperf/blob/f80980feaf5334532d21df621c5791202f19dfab/src/sockperf.cpp#L2556
* https://github.com/Mellanox/sockperf/blob/f80980feaf5334532d21df621c5791202f19dfab/src/client.h#L116
It's ironic that somewhere we have a comment "... to improve performance"

agalanin-at-nvidia replied:
`m_ServerList` map also copies `IPAddress` as a key. Probably we need to have three different maps for each socket address family (INET, INET6, UNIX) or variant type for map key that stores only 4 bytes for INET addresses etc.
https://github.com/Mellanox/sockperf/blob/f80980feaf5334532d21df621c5791202f19dfab/src/defs.h#L598
https://github.com/Mellanox/sockperf/blob/f80980feaf5334532d21df621c5791202f19dfab/src/client.h#L58

EldarShalev replied:
I see that the union is 
```
class IPAddress {
    union {
        in_addr m_addr4;
        in6_addr m_addr6;
    };
```

Should I add 
`const char *m_addr_un;`  
or 
`sockaddr_un *m_addr_un`

EldarShalev replied:
if it's marked as `const char` it can be added as 

```

  case AF_UNIX:
+            if (len >= sizeof(sockaddr_un)) {
+                m_addr_un = reinterpret_cast<const sockaddr_un *>(sa)->sun_path;
+            } else {
+                m_family = AF_UNSPEC;
+            }
+            break;

```

agalanin-at-nvidia replied:
Yes, idea with `char *` looks good. But not forget to strdup original `sun_path` on creation and free it in destructor.

agalanin-at-nvidia replied:
I thought more and realized that it is not so simple.

You must have full-length sockaddr for `recvfrom` because syscall known nothing about our optimizations. You also need to pass full sockaddr_un to `sendto` call. So I think that we must replace `sendto_addr = m_fds_ifd->server_addr;` with pointer assignment and move `char *` into `IPAddress` class.

agalanin-at-nvidia replied:
We still need to optimize these places:
* https://github.com/Mellanox/sockperf/blob/f80980feaf5334532d21df621c5791202f19dfab/src/server.h#L285
* https://github.com/Mellanox/sockperf/blob/f80980feaf5334532d21df621c5791202f19dfab/src/sockperf.cpp#L2556

EldarShalev replied:
How can we optimize this?

agalanin-at-nvidia replied:
At first, check that we can use pointer instead of copy. If it is not possible, copy only needed bytes depending on the address family.

EldarShalev replied:
Please take a loot at `copy_relevant_sockaddr_params` function in `common.h`

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -426,11 +426,11 @@ static int proc_mode_under_load(int id, int argc, const char **argv) {
           aopt_set_literal(0),
-          aopt_set_string("client_ip"),
-          "Force the client side to bind to a specific ip address (default = 0). " },
+          aopt_set_string("client_addr"),
```
```suggestion
          aopt_set_string("client_ip", "client_addr"),
```

EldarShalev replied:
@agalanin-at-nvidia  I found a bug in aopt_set_string. it doesn't work for more then one option. 
Reviewed this bug with @igor-ivanov.

agalanin-at-nvidia replied:
Fix (https://github.com/Mellanox/sockperf/pull/177) for this bug was already merged into master, so you can use multiple option names after rebase.

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -1718,7 +1718,16 @@ static int resolve_sockaddr(const char *host, const char *port, int sock_type,
+        ((sockaddr_store_t*)addr)->addr_un.sun_family = AF_UNIX;
+        ((sockaddr_store_t*)addr)->ss_family = AF_UNIX;
+
+        return 0;
```
```suggestion
        res = 0;
```
otherwise you will get memory leak at line 1731 (freeaddrinfo)

EldarShalev replied:
Will be fixed.

EldarShalev replied:
@agalanin-at-nvidia 
I am not sure it's correct because getaddrinfo was failed and the error that was returned is EAI_NONAME. 
using freeaddrinfo in that case will cause segmentation fault (attaching screenshot) 


![segmentation fault for freeaddrinfo](https://user-images.githubusercontent.com/44922030/164436342-a4acc3b5-cb71-439d-8d96-8c7ffedecc01.JPG)

agalanin-at-nvidia replied:
ok. please move freeaddinfo into `if (res == 0)` branch

EldarShalev replied:
created separate PR

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -1718,7 +1718,16 @@ static int resolve_sockaddr(const char *host, const char *port, int sock_type,
     int res = getaddrinfo(host, port, &hints, &result);
     if (res == 0) {
         get_socket_address(result, addr, addr_len);
+    } else if (res == EAI_NONAME && host != NULL) { // Unix Domain Socket name
```
This may be surprising if you have a host in your network with the same name as unix socket. I propose to add "unix:" prefix to UNIX socket names or force user to pass full path (starting with /).

EldarShalev replied:
Added "unix:" prefix as requested 
attaching screenshots for both cases if  --client_addr is provided or not 

![unix prefix with specific binding path](https://user-images.githubusercontent.com/44922030/167402290-4f90c6d1-33cc-477e-b9bd-bfa582bacdcf.JPG)
![unix prefix](https://user-images.githubusercontent.com/44922030/167402295-dfda5b81-3866-4ed5-b5e2-ebee377e00a8.JPG)

rosenbaumalex replied:
wait, I understand the request to have "unix:" be in the user interface for the --addr <string>
but remove it from the UDS sysfs file path

agalanin-at-nvidia replied:
@rosenbaumalex, yes I mean it.

I just realized that adding "unix:" will make feedfile parser more complex because `:` is a field separator. So I suggest to switch to another idea and force user to specify a full path to unix socket (starting with /).

@EldarShalev, please check that UNIX socket names is supported in feedfile parser (function `set_sockets_from_feedfile`).

igor-ivanov replied:
do you consider any special symbol at the start excluding `.:[0-9][a-f][A-F]` to simplify address ip4/ip6/unix recognition?

EldarShalev replied:
I thought about this regex for unix domain socket feedfile 
```
#define UNIX_DOMAIN_SOCKET_FORMAT_REG_EXP                                                          \
        "^[UuTt]:[Uu][Nn][Ii][Xx]\\/([0-9a-zA-z\\/]+)[\r\n]"

```

agalanin-at-nvidia replied:
I suggest `"^[UuTt]:[Uu][Nn][Ii][Xx]/(.+)[\\r\\n]"` because any symbol may be a part of socket name. Escaping is needed for `\r` and `\n` and not needed for `/`.

igor-ivanov replied:
I meant something as `~unix_socket_name` where `~` is the symbol that can not be a part of ipv4/ipv6 address as an alternative to `unix`

agalanin-at-nvidia replied:
`/` at the start is enough to distinguish between file name and ip address or host name

igor-ivanov replied:
I am fine with any symbol (it might be `/`) that allows to simplify ipv4/ipv6/unix recognition.

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3346,6 +3361,21 @@ int bringup(const int *p_daemonize) {
+                        if (strlen(s_user_params.client_bind_info.addr_un.sun_path) == 0 && s_user_params.mode == MODE_CLIENT) { // no specific addr client_info was provoided
+                            log_dbg("No client name was provided, setting addr_un.sun_path to %s_client\n",s_user_params.addr.addr_un.sun_path);
+                            char result[sizeof(s_user_params.addr.addr_un.sun_path) + sizeof("_client")];
+                            strcpy(result,s_user_params.addr.addr_un.sun_path);
```
you can use std::string here

EldarShalev replied:
Will be fixed

agalanin-at-nvidia reviewed (changes requested):
You also need to verify that UDS addresses were handled properly in the following places:
* Client::client_get_server_id (converted to IPAddress), used only in statistic mode
* sockaddr_to_hostport

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3367,10 +3367,9 @@ int bringup(const int *p_daemonize) {
-                            strcpy(s_user_params.client_bind_info.addr_un.sun_path, result);
+                            std::string result = s_user_params.addr.addr_un.sun_path;
+                            result += "_client";
+                            strcpy(s_user_params.client_bind_info.addr_un.sun_path, result.c_str());
```
Are you sure that you have enough memory in destination array?

EldarShalev replied:
sun_path size limit is 108 
https://man7.org/linux/man-pages/man7/unix.7.html

agalanin-at-nvidia replied:
108 + strlen("_client") > 108

igor-ivanov commented on doc/main.dox:
```diff
@@ -184,7 +184,7 @@ sockperf: [tid 4805] using epoll() to block on socket(s)
  -t      --time                 -Run for <sec> seconds (default 1, max = 36000000).
          --client_port          -Force the client side to bind to a specific port (default = 0).
-         --client_ip            -Force the client side to bind to a specific ip address (default = 0).
+         --client_addr          -Force the client side to bind to a specific address [IPV4/6, UDS_NAME] (default = 0).
```
please consider
`specific address in IPv4, IPv6, UNIX domain socket format (default = 0)`

**Renaming should be confirmed by verification!**

igor-ivanov commented on src/common.cpp:
```diff
@@ -55,11 +55,16 @@ std::string sockaddr_to_hostport(const struct sockaddr *addr)
     if (addr->sa_family == AF_INET6) {
         return "[" + std::string(hbuf) + "]:" + std::string(pbuf);
+    } else if (addr->sa_family == AF_UNIX ) {
+        return std::string(pbuf) + " [UDS]";
```
should not it be `std::string(hbuf)` only
see
man for `getnameinfo`
```
If	a link-layer address or	UNIX-domain address is passed to
     getnameinfo(), its	ASCII representation will be stored in host.  The
     string pointed to by serv will be set to the empty	string if non-NULL;
     flags will	always be ignored.
```

EldarShalev replied:
Not sure if I understand, currently I use 'pbuf' and it returns the path correctly
![pbuf](https://user-images.githubusercontent.com/44922030/165500906-315eac18-3293-4566-98e9-a895b6cd4ea5.JPG)

EldarShalev replied:
when using hbuf I see
![hbuf](https://user-images.githubusercontent.com/44922030/165503664-b5cddf3c-2975-43ef-a397-0bcd278675f1.JPG)

igor-ivanov replied:
ok

igor-ivanov commented on src/iohandlers.cpp:
```diff
@@ -36,9 +36,15 @@ static void print_addresses(const fds_data *data, int &list_count)
+            case AF_UNIX:
+                printf("[%2d] Address is %5s # UDS\n", list_count++, pbuf);
+                break;
+            default:
```
please consider using `PRINT_PROTOCOL()`
please check if you need `hbuf` instead of `pbuf` (see comment above)

EldarShalev replied:
Added PRINT_SOCKET_TYPE for clarifying that this is not protocol (TCP/UDP) but rather a socket type

igor-ivanov commented on src/sockperf.cpp:
```diff
@@ -165,7 +165,7 @@ static const AOPT_DESC common_opt_desc[] = {
     { OPT_TCP,                AOPT_NOARG,                       aopt_set_literal(0),
       aopt_set_string("tcp"), "Use TCP protocol (default UDP)." },
-    { 'i', AOPT_ARG, aopt_set_literal('i'), aopt_set_string("ip"), "Listen on/send to ip <ip>." },
+    { 'i', AOPT_ARG, aopt_set_literal('i'), aopt_set_string("addr", "ip"), "Listen on/send to ip <ip>, or address <name>" },
```
please consider
`Listen on/send to address in IPv4, IPv6, UNIX domain socket format`

igor-ivanov commented on src/sockperf.cpp:
```diff
@@ -426,11 +426,11 @@ static int proc_mode_under_load(int id, int argc, const char **argv) {
-          aopt_set_string("client_ip"),
-          "Force the client side to bind to a specific ip address (default = 0). " },
+          aopt_set_string("client_addr"),
+          "Force the client side to bind to a specific address [IPV4/6, UDS_NAME] (default = 0). " },
```
`.. address in IPv4, IPv6, UNIX domain socket format (default = 0)`

igor-ivanov commented on src/sockperf.cpp:
```diff
@@ -738,11 +738,11 @@ static int proc_mode_ping_pong(int id, int argc, const char **argv) {
-          aopt_set_string("client_ip"),
-          "Force the client side to bind to a specific ip address (default = 0). " },
+          aopt_set_string("client_addr"),
+          "Force the client side to bind to a specific address [IPV4/6, UDS_NAME] (default = 0). " },
```
`specific address in IPv4, IPv6, UNIX domain socket format (default = 0)`

igor-ivanov commented on src/sockperf.cpp:
```diff
@@ -1072,11 +1072,11 @@ static int proc_mode_throughput(int id, int argc, const char **argv) {
-          aopt_set_string("client_ip"),
-          "Force the client side to bind to a specific ip address (default = 0). " },
+          aopt_set_string("client_addr"),
+          "Force the client side to bind to a specific address [IPV4/6, UDS_NAME] (default = 0). " },
```
`specific address in IPv4, IPv6, UNIX domain socket format (default = 0)`

igor-ivanov commented on src/sockperf.cpp:
```diff
@@ -3346,6 +3360,20 @@ int bringup(const int *p_daemonize) {
             } else {
                 memset(reinterpret_cast<void *>(tmp), 0, sizeof(struct fds_data));
+                if (s_user_params.addr.ss_family == AF_UNIX) {
+                    log_dbg("UDS name was provided %s\n", s_user_params.addr.addr_un.sun_path);
```
`/UDS/UNIX domain socket/`

igor-ivanov replied:
can you consider using `UNIX domain socket` instead of `UDS`

EldarShalev replied:
I changed it in latest commit, this is 'Outdated'

agalanin-at-nvidia commented on doc/main.dox:
```diff
@@ -184,7 +184,7 @@ sockperf: [tid 4805] using epoll() to block on socket(s)
  -t      --time                 -Run for <sec> seconds (default 1, max = 36000000).
          --client_port          -Force the client side to bind to a specific port (default = 0).
-         --client_ip            -Force the client side to bind to a specific ip address (default = 0).
+         --client_addr          -Force the client side to bind to a specific address in IPv4, IPv6, UNIX domaon socket format (default = 0).
```
```suggestion
         --client_addr          -Force the client side to bind to a specific address in IPv4, IPv6, UNIX domain socket format (default = 0).
```

agalanin-at-nvidia commented on src/common.cpp:
```diff
@@ -55,11 +55,16 @@ std::string sockaddr_to_hostport(const struct sockaddr *addr)
     if (addr->sa_family == AF_INET6) {
         return "[" + std::string(hbuf) + "]:" + std::string(pbuf);
+    } else if (addr->sa_family == AF_UNIX) {
+        return std::string(pbuf) + " [UDS]";
```
`[UNIX]` is more convenient

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3346,6 +3360,24 @@ int bringup(const int *p_daemonize) {
+                            std::string result = s_user_params.addr.addr_un.sun_path;
+                            result += "_client";
+                            if (strlen(result.c_str()) > MAX_UDS_NAME) {
+                                log_err("length of name is larger then %d bytes", MAX_UDS_NAME);
```
```suggestion
                                log_err("length of name is larger than %d bytes", MAX_UDS_NAME);
```

EldarShalev committed:
issue 3109144: Supporting feedfile for unix domain socket

Signed-off-by: Eldar Shalev <eldarsh@nvidia.com>
Adding new support for feedfile to include Unix Domain Socket on 2 modes - SOCK_STREAM and SOCK_DGRAM.
Binding automatically to the given path as following: <PATH>_<PID>_<FD>.
    Example:
	u:/tmp/test  -> will create a unix domain socket of DGRAM and path '/tmp/test'
	t:/tmp/test2 -> will create a unix domain socket of STREAM and path '/tmp/test2'

* path must starts with '/'

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -1718,7 +1718,22 @@ static int resolve_sockaddr(const char *host, const char *port, int sock_type,
+        log_dbg("provided path for Unix Domain Socket is %s\n", host);
+        std::string result = "unix:";
+        result += host;
+        if (strlen(result.c_str()) > MAX_UDS_NAME) {
```
```suggestion
        if (result.length() > MAX_UDS_NAME) {
```
also in line 1729

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3346,6 +3366,24 @@ int bringup(const int *p_daemonize) {
+                            log_dbg("No client name was provided, setting addr_un.sun_path to %s_client\n",s_user_params.addr.addr_un.sun_path);
+                            std::string result = s_user_params.addr.addr_un.sun_path;
+                            result += "_client";
+                            if (strlen(result.c_str()) > MAX_UDS_NAME) {
```
```suggestion
                            if (result.length() > MAX_UDS_NAME) {
```

agalanin-at-nvidia commented on src/ip_address.h:
```diff
@@ -83,6 +86,13 @@ class IPAddress {
             break;
+        case AF_UNIX:
+            if (len >= sizeof(sockaddr_un)) {
+                m_addr_un = reinterpret_cast<const sockaddr_un *>(sa)->sun_path;
```
must be copied because `sa` may be destroyed separately

agalanin-at-nvidia commented on src/ip_address.h:
```diff
@@ -115,6 +130,8 @@ class IPAddress {
             case AF_INET6:
                 return IN6_ARE_ADDR_EQUAL(&key1.m_addr6, &key2.m_addr6);
+            case AF_UNIX:
+                return key1.m_addr_un == key2.m_addr_un;
```
`strncmp(key1.m_addr_un, key2.m_addr_un, MAX_UNIX_LEN) == 0` or use `std::string` to store values

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -1718,7 +1718,27 @@ static int resolve_sockaddr(const char *host, const char *port, int sock_type,
+    } else if (res == EAI_NONAME && host != NULL) { // Unix Domain Socket name
+        log_dbg("provided path for Unix Domain Socket is %s\n", host);
+        std::string path = host;
+        if (!(path.rfind("/", 0) == 0)) { // limit to prefix
```
1. `path.size() > 0 && path[0] == '/'`
2. It's better to move this check before `getaddrinfo` invocation because getaddrinfo on file name will always fail.

EldarShalev replied:
1) will be changed 
2) if I move the condition before getaddrinfo, so in case we have path "tmp/test" and not "/tmp/test" 
it will not enter to first condition - since it doesn't start with '/', and will fail on the second.

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -1718,7 +1718,27 @@ static int resolve_sockaddr(const char *host, const char *port, int sock_type,
+            res = SOCKPERF_ERR_SOCKET;
+            return res;
+        }
+        addr_len = path.length() + 2;
+        strncpy(((sockaddr_store_t*)addr)->addr_un.sun_path, path.c_str(), addr_len);
```
1. I don't understand why you adding 2 bytes to string length.
2. It's enough to call `strncpy(, , MAX_UDS_NAME)`. This handles corner case when `path.length() == MAX_UDS_NAME` so trailing zero is not needed to be copied.

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3032,7 +3057,7 @@ static int set_sockets_from_feedfile(const char *feedfile_name) {
                 is_unspec_addr(s_user_params.client_bind_info)) {
                 log_msg("socketxtreme requires forcing the client side to bind to a specific ip "
-                        "address (use --client_ip) option");
+                        "address (use --client_addr) option");
```
```suggestion
                        "address (use --client_ip/--client_addr option)");
```

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3346,6 +3371,24 @@ int bringup(const int *p_daemonize) {
+                                log_err("length of name is larger than %d bytes", MAX_UDS_NAME);
+                                rc = SOCKPERF_ERR_SOCKET;
+                            } else
+                                strcpy(s_user_params.client_bind_info.addr_un.sun_path, sun_path.c_str());
```
`strncpy(, , MAX_UDS_NAME)`

EldarShalev replied:
also removed in 'feedfile' commit

agalanin-at-nvidia commented on src/ip_address.h:
```diff
@@ -101,6 +111,11 @@ class IPAddress {
 
+    inline const char *addr_un() const
+    {
+        return m_addr_un;
```
It's dangerous because `m_addr_un` may not contain trailing zero.

EldarShalev replied:
Not sure how to approach this issue

agalanin-at-nvidia replied:
The simplest way is to use `std::string` as a storage.

EldarShalev replied:
string * with new/delete
or char* with strdup and free?

agalanin-at-nvidia replied:
I propose to use just a member of type std::string outside the `union`. This will slightly increase structure size but not so dramatically as for char[MAX_UDS_NAME].

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3346,6 +3371,24 @@ int bringup(const int *p_daemonize) {
+                    s_user_params.tcp_nodelay = false;
+                    if (s_user_params.sock_type == SOCK_DGRAM) { // Need to bind localy
+                        s_user_params.client_bind_info.ss_family = AF_UNIX;
+                        if (strlen(s_user_params.client_bind_info.addr_un.sun_path) == 0 && s_user_params.mode == MODE_CLIENT) { // no specific addr client_info was provoided
```
```suggestion
                        if (s_user_params.client_bind_info.addr_un.sun_path[0] == 0 && s_user_params.mode == MODE_CLIENT) { // no specific addr client_info was provoided
```

EldarShalev replied:
"_client" was replaced in 'feedfile' commit for
 PATH_PID_FD

agalanin-at-nvidia replied:
my comment about strlen is still actual

EldarShalev replied:
we can close this one, changed from `strlen` to `[0]`

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -868,6 +868,31 @@ int Client<IoType, SwitchDataIntegrity, SwitchActivityInfo, SwitchCycleDuration,
+                    p_client_bind_addr = &unix_addr;
+                    client_bind_addr_len = unix_addr_len;
+                    }
+                }
```
extract it to a separate function

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -868,6 +868,31 @@ int Client<IoType, SwitchDataIntegrity, SwitchActivityInfo, SwitchCycleDuration,
+                        char ifd_buffer [MAX_FDS_NUM];
+                        snprintf(pid_buffer, sizeof(pid_buffer), "%d", getpid());
+                        snprintf(ifd_buffer, sizeof(ifd_buffer), "%d", ifd);
+                        sun_path = sun_path + "_" + pid_buffer + "_" + ifd_buffer;
```
I suggest to do something like this:
```suggestion
                        std::string sun_path = sun_path_to_string(s_user_params.addr.addr_un.sun_path) + '_' + std::to_string(getpid()) + '_' + std::to_string(ifd);
```

You can start to use `std::to_string` because C++11 support is coming in PR #185.

`sun_path_to_string` must handle the corner case when sun_path does not include the trailing zero.

agalanin-at-nvidia replied:
Your new function `sun_path_to_string` does not match the name (because does not perform conversion from sun_path to std_string inside the function) and new code still lacks handling of absent zero at the end of buffer.

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -868,6 +868,31 @@ int Client<IoType, SwitchDataIntegrity, SwitchActivityInfo, SwitchCycleDuration,
+                        }
+                    p_client_bind_addr = &unix_addr;
+                    client_bind_addr_len = unix_addr_len;
+                    }
```
something is bad with indentation

agalanin-at-nvidia commented on src/defs.h:
```diff
@@ -154,8 +156,14 @@ const uint32_t TEST_FIRST_CONNECTION_FIRST_PACKET_TTL_THRESHOLD_MSEC = 50;
     "(:([a-zA-Z0-9\\.\\-]+|\\[([a-fA-F0-9.:]+)\\]))?[\r\n]"
+
+#define UNIX_DOMAIN_SOCKET_FORMAT_REG_EXP                                                          \
+        "^[UuTt]:[\\/]([0-9a-zA-z\\/]+)[\r\n]"
```
```suggestion
        "^[UuTt]:(/.+)[\r\n]"
```
Are you sure that `/` really needs escaping?

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -1713,12 +1713,33 @@ static int resolve_sockaddr(const char *host, const char *port, int sock_type,
+    int res;
     struct addrinfo *result;
-    int res = getaddrinfo(host, port, &hints, &result);
+    std::string path = host;
```
What if host == NULL?

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -1713,12 +1713,33 @@ static int resolve_sockaddr(const char *host, const char *port, int sock_type,
+    res = getaddrinfo(host, port, &hints, &result);
     if (res == 0) {
         get_socket_address(result, addr, addr_len);
+    } else if (res == EAI_NONAME && host != NULL) { // unix case but didn't start with '/'
+        log_err("unix domain socket path have to start with '/'");
+        res = SOCKPERF_ERR_SOCKET;
+        return res;
```
This also can happen when DNS server is not available, so it will be strange to get message about UNIX socket in this case. I think that it's better to just return an error without any hints.

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -2975,10 +3012,19 @@ static int set_sockets_from_feedfile(const char *feedfile_name) {
-    if (regexp_error != 0) {
-        log_msg("Failed to compile regexp");
+
+    // 1) if the line contains UNIX/unix in any way it's unix_domain socket
+    // 2) else it's ip format
+    //const char *format = (s_user_params.addr.ss_family == AF_UNIX ? UNIX_DOMAIN_SOCKET_FORMAT_REG_EXP : IP_PORT_FORMAT_REG_EXP);
```
obsolete comment

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3012,13 +3058,20 @@ static int set_sockets_from_feedfile(const char *feedfile_name) {
+                addr.erase(std::remove(addr.begin(), addr.end(), '\n'), addr.end());
+            } else {
+                log_msg("Invalid input in line %s: "
+                        "each line must have the following format: ip:port or type:ip:port or "
+                        "type:ip:port:mc_src_ip or unix/path_to_unix_socket. "
+                        "IPv6 addresses must be enclosed in square brackets.",
```
add line format example into function comment and README file

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3012,13 +3058,20 @@ static int set_sockets_from_feedfile(const char *feedfile_name) {
+            regexpres = regexec(&regexpr_unix, line, NUM_RE_GROUPS, m, 0);
+            if (regexpres == 0) {
+                type = line[0];
+                addr = line + m[1].rm_so - 1;
+                addr.erase(std::remove(addr.begin(), addr.end(), '\n'), addr.end());
```
After updating regexp (see my comment for defs.h) this can be simplified
```suggestion
                addr = std::string(line + m[1].rm_so, line + m[1].rm_eo);;
```

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3096,6 +3149,13 @@ static int set_sockets_from_feedfile(const char *feedfile_name) {
 
+            if (tmp->server_addr.ss_family == AF_UNIX) {
+                s_user_params.tcp_nodelay = false;
+                strcpy(s_user_params.addr.addr_un.sun_path,tmp->server_addr.addr_un.sun_path);
+                s_user_params.addr.ss_family = AF_UNIX;
```
why not copy the whole `tmp->server_addr`?

EldarShalev replied:
agree, `tmp->server_addr` is copied

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -868,6 +872,27 @@ int Client<IoType, SwitchDataIntegrity, SwitchActivityInfo, SwitchCycleDuration,
+                socklen_t unix_addr_len;
+                if (p_client_bind_addr->ss_family == AF_UNIX && g_pApp->m_const_params.sock_type == SOCK_DGRAM) { // Need to bind localy
+                    if (strlen(p_client_bind_addr->addr_un.sun_path) == 0) { // no specific addr client_info was provoided
+                        log_dbg("No client name was provided, setting addr_un.sun_path to %s_%d_%d\n",
+                                s_user_params.addr.addr_un.sun_path, getpid(),ifd);
```
you can use `%s` and `sun_path.c_str()` here instead of duplicating string concatenation here.

EldarShalev replied:
Are you referring to the  `sun_path_to_string` func or to the `log_dbg` print?

agalanin-at-nvidia replied:
I propose something like this code:
```C++

std::string sun_path = sun_path_to_string(s_user_params.addr.addr_un.sun_path) + '_' + std::to_string(getpid()) + '_' + std::to_string(ifd);
log_dbg("No client name was provided, setting addr_un.sun_path to %s\n",
        sun_path.c_str());
```

agalanin-at-nvidia commented on src/ip_address.h:
```diff
@@ -41,12 +43,15 @@
 #include <unordered_map>
 #endif
 
+#define MAX_UDS_NAME 108
```
This macro is defined also in `defs.h`. Please use only a single definition.

EldarShalev replied:
but `ip_address.h` doesn't include `defs.h` and I can't extern it either

agalanin-at-nvidia replied:
You can either include `defs.h` or extract `constants.h` from it.

EldarShalev replied:
`defs.h` already includes `ip_address.h`

EldarShalev replied:
So I'll leave single definition in `ip_address.h` only

agalanin-at-nvidia commented on src/ip_address.h:
```diff
@@ -83,6 +88,13 @@ class IPAddress {
             break;
+        case AF_UNIX:
+            if (len >= sizeof(sockaddr_un)) {
+                strncpy(m_addr_un ,reinterpret_cast<const sockaddr_un *>(sa)->sun_path, MAX_UDS_NAME);
```
bad formatting

EldarShalev replied:
where exactly?
![format issue](https://user-images.githubusercontent.com/44922030/169782735-c53256e2-9a98-4bf1-81ff-a8330f79f986.JPG)

agalanin-at-nvidia replied:
`m_addr_un ,reinterpret_cast`

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -2304,6 +2323,21 @@ void cleanup() {
                 }
+                if (s_user_params.addr.ss_family == AF_UNIX) {
+                    unlink(s_user_params.client_bind_info.addr_un.sun_path);
+                    if (s_user_params.mode == MODE_CLIENT && s_user_params.sock_type == SOCK_DGRAM) { // unlink binded client
+                        std::string sun_path = s_user_params.addr.addr_un.sun_path;
+                        char pid_buffer [MAX_FDS_NUM];
+                        char ifd_buffer [MAX_FDS_NUM];
+                        snprintf(pid_buffer, sizeof(pid_buffer), "%d", getpid());
+                        snprintf(ifd_buffer, sizeof(ifd_buffer), "%d", ifd);
+                        sun_path = sun_path + "_" + pid_buffer + "_" + ifd_buffer;
```
You're using similar code in `Client::initBeforeLoop()`. Please extract to a common function.

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -2304,6 +2323,21 @@ void cleanup() {
+                        unlink(sun_path.c_str());
+                    }
+                    if (s_user_params.mode == MODE_SERVER)
+                        unlink(g_fds_array[ifd]->server_addr.addr_un.sun_path);
```
Will cause out-of-buffer-read if socket name is `/tmp/socket00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000` (MAX_UDS_NAME symbols long)

EldarShalev replied:
but wouldn't we fail much earilier in `resolve_sockaddr` stage, when there is a check if 
`path.length() > MAX_UDS_NAME`?

agalanin-at-nvidia replied:
I'm talking about `path.length() == MAX_UDS_NAME`. Probably the best solution is to check length **including** trailing zero and force the last symbol in buffer to be zero.

EldarShalev replied:
Why not just change the condition from `path.length() > MAX_UDS_NAME` to `path.length() >= MAX_UDS_NAME`?

agalanin-at-nvidia replied:
Yes, that's exactly what I mean.

EldarShalev replied:
can we close this one?

agalanin-at-nvidia commented on src/common.h:
```diff
@@ -182,4 +182,13 @@ static inline void sockaddr_set_portn(sockaddr_store_t &addr, uint16_t port)
 
+/**
+ * @brief
+ * @return concatenation of unix domaon socket string, useful for multiple binding.
+ */
+static inline std::string sun_path_to_string(std::string sun_path, std::string pid, std::string ifd) {
```
1. Function content does not match the name: sun_path to string conversion is actually performed outside this function. Actual name for this function is build_client_address.
2. Use `const std::string &` to avoid unnecessary copies.
3. Pass `pid` and `ifd` as numbers and convert them inside a function.
4. Remove empty line.

agalanin-at-nvidia replied:
item 1 (Function content does not match the name) is still actual

agalanin-at-nvidia commented on src/ip_address.h:
```diff
@@ -79,6 +84,13 @@ class IPAddress {
             break;
+        case AF_UNIX:
+            if (len >= sizeof(sockaddr_un)) {
+                strncpy(m_addr_un, reinterpret_cast<const sockaddr_un *>(sa)->sun_path, MAX_UDS_NAME);
```
Destination memory must be allocated before copy.

EldarShalev committed:
issue 3109144: Initialize default values

agalanin-at-nvidia commented on src/defs.h:
```diff
@@ -631,80 +631,85 @@ typedef enum { // must be coordinated with s_fds_handle_desc in common.cpp
+    bool b_server_dont_reply = false;
+    bool b_server_detect_gaps = false;
+    uint32_t mps = MPS_DEFAULT; // client side only
     struct sockaddr_store_t client_bind_info;
```
= nullptr;

EldarShalev replied:
but it's not a pointer

agalanin-at-nvidia replied:
It was a comment for `select_timeout`. I misplaced the mouse.

agalanin-at-nvidia commented on src/defs.h:
```diff
@@ -631,80 +631,85 @@ typedef enum { // must be coordinated with s_fds_handle_desc in common.cpp
+    bool withsock_accl = false;
+    int max_looping_over_recv = 1;
+    int tos = 0x00;
     unsigned int lls_usecs;
```
= 0;

agalanin-at-nvidia commented on src/defs.h:
```diff
@@ -631,80 +631,85 @@ typedef enum { // must be coordinated with s_fds_handle_desc in common.cpp
+    int max_looping_over_recv = 1;
+    int tos = 0x00;
     unsigned int lls_usecs;
     bool lls_is_set;
```
= false;

agalanin-at-nvidia commented on src/defs.h:
```diff
@@ -631,80 +631,85 @@ typedef enum { // must be coordinated with s_fds_handle_desc in common.cpp
-    uint32_t dummy_mps;                   // client side only
+    uint32_t dummy_mps = 0;                   // client side only
     TicksDuration dummySendCycleDuration; // client side only
     uint32_t rate_limit;
```
= 0;

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -2412,76 +2412,23 @@ void set_defaults() {
-    s_user_params.histogram_lower_range = 0;
-    s_user_params.histogram_upper_range = 2000000;
-    s_user_params.histogram_bin_size = 10;
+    g_debug_level = LOG_LVL_INFO;
```
what is the reason for this change?

EldarShalev replied:
If we set default values at `defs.h`, do we need it also in `set_default` func? 
I removed the default values there

agalanin-at-nvidia replied:
I'm talking about line ` g_debug_level = LOG_LVL_INFO;` that does not present in previous version of this file.

EldarShalev replied:
https://github.com/Mellanox/sockperf/blob/sockperf_v2/src/sockperf.cpp

![log debug](https://user-images.githubusercontent.com/44922030/171401829-2c01307a-ae85-43ef-9ac1-331ced5b1c2c.JPG)

agalanin-at-nvidia replied:
ok, now I see it

agalanin-at-nvidia commented on src/defs.h:
```diff
@@ -631,80 +631,85 @@ typedef enum { // must be coordinated with s_fds_handle_desc in common.cpp
+    unsigned int packetrate_stats_print_ratio = 0;
+    unsigned int burst_size = 1;
+    bool packetrate_stats_print_details = false;
     //	bool stream_mode; - use b_stream instead
```
can be removed

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -879,6 +879,26 @@ int Client<IoType, SwitchDataIntegrity, SwitchActivityInfo, SwitchCycleDuration,
+                            log_err("length of name is larger than %d bytes", MAX_UDS_NAME);
+                            rc = SOCKPERF_ERR_SOCKET;
+                        } else {
+                            strncpy(unix_addr.addr_un.sun_path, sun_path.c_str(), MAX_UDS_NAME);
```
it is better to memset `unix_addr.addr_un.sun_path` with zeroes before this to strip all the garbage at the end of the buffer

EldarShalev replied:
I can memset in `resolve_sockaddr` func only before copying

agalanin-at-nvidia replied:
You need to call it here to fill destination array.

EldarShalev replied:
When I tried to memset `s_user_params.addr.addr_un.sun_path` here it erased the path
do you mean to memset  local `std::string sun_path`?

agalanin-at-nvidia replied:
Call memset before strncpy.

EldarShalev replied:
Ok yes, I mistakenly looked at another path

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -879,6 +879,26 @@ int Client<IoType, SwitchDataIntegrity, SwitchActivityInfo, SwitchCycleDuration,
+                            rc = SOCKPERF_ERR_SOCKET;
+                        } else {
+                            strncpy(unix_addr.addr_un.sun_path, sun_path.c_str(), MAX_UDS_NAME);
+                            unix_addr_len = sizeof(struct sockaddr_un);
```
`sizeof(struct sockaddr_un) > sizeof(struct sockaddr_store_t)` so we're not correct here. I propose to use `sockaddr_un` as the storage for `unix_addr`.

EldarShalev replied:
Not sure which parameter you're referring to

agalanin-at-nvidia replied:
OK, now I see that `sizeof(struct sockaddr_un) == sizeof(struct sockaddr_store_t)`.

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -998,6 +1018,7 @@ int Client<IoType, SwitchDataIntegrity, SwitchActivityInfo, SwitchCycleDuration,
     return rc;
 }
 
+
```
extra empty line

agalanin-at-nvidia commented on src/defs.h:
```diff
@@ -624,80 +631,85 @@ typedef enum { // must be coordinated with s_fds_handle_desc in common.cpp
     uint32_t cooldown_msec;
     uint32_t warmup_msec;
     uint64_t cooldown_num;
     uint64_t warmup_num;
```
= 0;

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -2383,76 +2412,23 @@ void set_defaults() {
-    s_user_params.b_client_ping_pong = false;
-    s_user_params.b_no_rdtsc = false;
     memset(s_user_params.sender_affinity, 0, sizeof(s_user_params.sender_affinity));
     memset(s_user_params.receiver_affinity, 0, sizeof(s_user_params.receiver_affinity));
```
I propose to move these lines into class constructor

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -2948,11 +2924,15 @@ static int set_sockets_from_feedfile(const char *feedfile_name) {
 #endif // NEED_REGEX_WORKAROUND
 
+
+
```
extra empty lines

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3038,13 +3024,20 @@ static int set_sockets_from_feedfile(const char *feedfile_name) {
-                    line);
-            rc = SOCKPERF_ERR_INCORRECT;
-            break;
+            regexpres = regexec(&regexpr_unix, line, NUM_RE_GROUPS, m, 0);
+            if (regexpres == 0) {
+                type = line[0];
+                addr = std::string(line + m[1].rm_so, line + m[1].rm_eo);
+            } else {
+                log_msg("Invalid input in line %s: "
+                        "each line must have the following format: ip:port or type:ip:port or "
+                        "type:ip:port:mc_src_ip or type:/PATH. "
+                        "IPv6 addresses must be enclosed in square brackets."
+                        "Unix domain socket must start with '/'",
+                        line);
+                rc = SOCKPERF_ERR_INCORRECT;
+                break;
+            }
```
Your code will not compile with something other than gcc 4.8.5.

You can reproduce compilation error on gcc 4.8.5 by commenting `#define NEED_REGEX_WORKAROUND` at the start of this file.

EldarShalev replied:
not sure I understand. 
I saw the new `#define NEED_REGEX_WORKAROUND`
what is the issue?

agalanin-at-nvidia replied:
Code must compile and work for gcc 4.8.5 (in this case NEED_REGEX_WORKAROUND is defined) and for another compilers (when NEED_REGEX_WORKAROUND is not defined). Since gcc 4.8.5 has regex error only in runtime you can check *compilation* by temporary commenting `#define NEED_REGEX_WORKAROUND`.

EldarShalev replied:
all the "unix regex" should be under NEED_REGEX_WORKAROUND definition?

agalanin-at-nvidia replied:
Both regexps must be used in both modes. This is just a workaround to support old compiler.

agalanin-at-nvidia replied:
Check compilation logs ("Mellanox Lab" build below) to see how your code compiled using different compilers.

EldarShalev replied:
Do we want to support new UDS feature with feedfile for old compiler?

agalanin-at-nvidia replied:
yes

agalanin-at-nvidia replied:
reminder

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3114,6 +3107,13 @@ static int set_sockets_from_feedfile(const char *feedfile_name) {
 
+        if (tmp->server_addr.ss_family == AF_UNIX) {
+            s_user_params.tcp_nodelay = false;
+            strcpy(s_user_params.addr.addr_un.sun_path,tmp->server_addr.addr_un.sun_path);
```
It's hard to me to understand where we perform `sun_path` length validation. Is string length in `tmp` already validated?

EldarShalev replied:
AFAIK when we called `resolve_addr`

EldarShalev replied:
`resolve_sockaddr` func validate if the length is less than or eqaul to MAX_UDS_NAME

agalanin-at-nvidia replied:
OK, I see.

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3373,19 +3384,19 @@ int bringup(const int *p_daemonize) {
+                    /* create a socket */
                 if ((curr_fd = (int)socket(tmp->server_addr.ss_family, tmp->sock_type, 0)) <
                     0) { // TODO: use SOCKET all over the way and avoid this cast
-                    log_err("socket(AF_INET4/6, SOCK_x)");
+                    log_err("socket(AF_INET4/6 /AF_UNIX, SOCK_x)");
                     rc = SOCKPERF_ERR_SOCKET;
                 } else {
                     if ((curr_fd >= MAX_FDS_NUM) ||
-                        (prepare_socket(curr_fd, tmp.get()) ==
-                         (int)INVALID_SOCKET)) { // TODO: use SOCKET all over the way and avoid
-                                                 // this cast
-                        log_err("Invalid socket");
-                        close(curr_fd);
-                        rc = SOCKPERF_ERR_SOCKET;
+                    (prepare_socket(curr_fd, tmp.get()) ==
+                    (int)INVALID_SOCKET)) { // TODO: use SOCKET all over the way and avoid
+                                            // this cast
+                    log_err("Invalid socket");
+                    close(curr_fd);
+                    rc = SOCKPERF_ERR_SOCKET;
```
looks like this code isn't properly formatted

EldarShalev replied:
Done.

agalanin-at-nvidia commented:
Please rebase your branch to resolve merge conflicts in `sockperf.cpp` reported by GitHub.

EldarShalev commented:
> 

Which conflicts?

agalanin-at-nvidia commented:
> Which conflicts?

![image](https://user-images.githubusercontent.com/91539962/171584574-3806a8e6-d19e-4bdd-92f8-fb5853eed9f5.png)

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -879,6 +879,27 @@ int Client<IoType, SwitchDataIntegrity, SwitchActivityInfo, SwitchCycleDuration,
+                struct sockaddr_store_t unix_addr;
+                socklen_t unix_addr_len;
+                if (p_client_bind_addr->ss_family == AF_UNIX && g_pApp->m_const_params.sock_type == SOCK_DGRAM) { // Need to bind localy
+                    if (strlen(p_client_bind_addr->addr_un.sun_path) == 0) { // no specific addr client_info was provoided
```
use `[0] == 0` instead of `strlen` as discussed before

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -879,6 +879,27 @@ int Client<IoType, SwitchDataIntegrity, SwitchActivityInfo, SwitchCycleDuration,
+                        log_dbg("No client name was provided, setting addr_un.sun_path to %s\n",
+                               sun_path.c_str());
+                        if (sun_path.length() > MAX_UDS_NAME) {
+                            log_err("length of name is larger than %d bytes", MAX_UDS_NAME);
```
```suggestion
                            log_err("length of client socket name (%s) is more than %d bytes", sun_path.c_str(), MAX_UDS_NAME);
```

agalanin-at-nvidia commented on src/iohandlers.cpp:
```diff
@@ -36,9 +36,15 @@ static void print_addresses(const fds_data *data, int &list_count)
+            default:
+                printf("[%2d] IP = %-15s PORT = %5s # %s\n", list_count++,
                 hbuf, pbuf,
                 PRINT_PROTOCOL(data->sock_type));
```
bad formatting

agalanin-at-nvidia commented on src/ip_address.h:
```diff
@@ -79,6 +84,13 @@ class IPAddress {
             break;
+        case AF_UNIX:
+            if (len >= sizeof(sockaddr_un)) {
+                m_addr_un = std::string(reinterpret_cast<const sockaddr_un *>(sa)->sun_path);
```
`sun_path` of remote peer comes from uncontrolled source in statistic mode. We must carefully check it.

Once again I propose to create a common function `sun_path_to_string` that handles a corner case when `sun_path` array does not contain a trailing zero byte.

EldarShalev replied:
I agree but I can't put it in `common.h ` because `common.h` -> `defs.h` -> `ip_address.h`

agalanin-at-nvidia replied:
You can use forward declaration. It's not too beautiful but can be a solution in this case.

EldarShalev replied:
In this corner case when  `sun_path` array does not contain a trailing zero - 
should I manually add it or fail the function?

agalanin-at-nvidia replied:
It's a valid name for remote peer so just add one zero byte to the end.

Note that there are *abstract* UNIX sockets whose name is started with zero byte. This case must be handled separately (see https://man7.org/linux/man-pages/man7/unix.7.html -> address format -> abstract).

Commonly used programs such as `ss` or `lsof` displays abstract UNIX socket names with `@` instead of first zero byte. This may be useful in debug print and statistics mode.

EldarShalev replied:
AFAIK, The thing is with abstract socket, null bytes in `sun_path` have no special significance, so they are not acting as a termination character.

The only way for `bind `function to know what's the length of `sun_path `when using abstract socket, is with its `addrlen` argument.

if `addrlen` is set to `sizeof(struct sockaddr_un)`, bind thinks that `sun_path` length is 108 (`sockaddr_un `implementation of `sun_path` is char `sun_path[108]`) and creates a socket with `@NAME[...](null bytes)`

Maybe in Abstract sockets cases, the correct way of setting `addr_len` is to set it to `sizeof(sun_family)` + the actual length of path (`strlen("#/tmp/test")` for example). This way, the socket will be name "@/tmp/test", without the null byte at the end.

Didn't test it thought..

EldarShalev replied:
Maybe can we deal with Abstract sockets in another phase? 
I am not sure how broad this case and if it justifies the work.

agalanin-at-nvidia replied:
yes, I think that we can postpone this case

agalanin-at-nvidia commented on src/ip_address.h:
```diff
@@ -79,6 +84,13 @@ class IPAddress {
                 m_family = AF_UNSPEC;
             }
             break;
+        case AF_UNIX:
```
This constructor is a big enough to move it into .cpp file.

EldarShalev replied:
Is it ok now?

agalanin-at-nvidia commented on src/ip_address.h:
```diff
@@ -111,10 +128,14 @@ class IPAddress {
         }
         return false;
     }
+
```
extra empty line

agalanin-at-nvidia commented on src/server.cpp:
```diff
@@ -254,7 +254,7 @@ int Server<IoType, SwitchActivityInfo, SwitchCalcGaps>::server_accept(int ifd) {
         fds_data *tmp;
 
-        tmp = (struct fds_data *)MALLOC(sizeof(struct fds_data));
+        tmp = new fds_data;
```
1. `tmp` must be deallocated using the same method (using `delete` instead of `FREE()`). Or it is better to use `std::unique_ptr` to simplify error case handling.
2. `if (!tmp)` is always false for throwing `new` operator.

agalanin-at-nvidia replied:
reminder

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -2310,6 +2328,16 @@ void cleanup() {
+                    unlink(s_user_params.client_bind_info.addr_un.sun_path);
+                    if (s_user_params.mode == MODE_CLIENT && s_user_params.sock_type == SOCK_DGRAM) { // unlink binded client
+                        std::string sun_path = sun_path_to_string(s_user_params.addr.addr_un.sun_path, getpid(), ifd);
+                        log_dbg("unlinking %s",sun_path.c_str());
```
```suggestion
                        log_dbg("unlinking %s", sun_path.c_str());
```

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -2383,76 +2411,21 @@ void set_defaults() {
-    s_user_params.b_no_rdtsc = false;
-    memset(s_user_params.sender_affinity, 0, sizeof(s_user_params.sender_affinity));
-    memset(s_user_params.receiver_affinity, 0, sizeof(s_user_params.receiver_affinity));
     // s_user_params.b_load_vma = false;
```
remove it

agalanin-at-nvidia replied:
reminder

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3114,6 +3102,13 @@ static int set_sockets_from_feedfile(const char *feedfile_name) {
 
+        if (tmp->server_addr.ss_family == AF_UNIX) {
+            s_user_params.tcp_nodelay = false;
+            strcpy(s_user_params.addr.addr_un.sun_path,tmp->server_addr.addr_un.sun_path);
```
```suggestion
            strcpy(s_user_params.addr.addr_un.sun_path, tmp->server_addr.addr_un.sun_path);
```

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3373,19 +3379,19 @@ int bringup(const int *p_daemonize) {
                 rc = SOCKPERF_ERR_NO_MEMORY;
             } else {
-                /* create a socket */
+                    /* create a socket */
```
Is this line properly formatted?

EldarShalev replied:
Seems ok in my IDE
![format bringup](https://user-images.githubusercontent.com/44922030/172371611-cc88a192-7d53-4056-9261-61001d57055a.JPG)

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3413,6 +3419,7 @@ int bringup(const int *p_daemonize) {
                 }
             }
 
+
```
extra empty line

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -1725,9 +1725,27 @@ static int resolve_sockaddr(const char *host, const char *port, int sock_type,
+                return res;
+            }
+            addr_len = MAX_UDS_NAME;
+            strncpy(((sockaddr_store_t*)addr)->addr_un.sun_path, path.c_str(), addr_len);
```
Please extract `(sockaddr_store_t*)addr` as a local variable to make code more clean.

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -1725,9 +1725,27 @@ static int resolve_sockaddr(const char *host, const char *port, int sock_type,
+                res = SOCKPERF_ERR_SOCKET;
+                return res;
+            }
+            addr_len = MAX_UDS_NAME;
```
UNIX socket address contains not only a path but also a `sun_family` member. So `addr_len` must be set to `sizeof(sockaddr_store_t)`.

EldarShalev replied:
When changing `addr_len = MAX_UDS_NAME` to `addr_len = sizeof(sockaddr_store_t);`
I get an issue: 
` ERROR: [fd=3] Cant bind socket, IP to bind: /tmp/test [UNIX]`

agalanin-at-nvidia replied:
It's probably because you're using `addr_len` in `strncpy` call one line below. Third argument for `strncpy` must be really `MAX_UDS_NAME`. But `addrlen` is an output parameter that must be set accordingly to socket address usage rules (to `sizeof(sockaddr_store_t)`).

EldarShalev replied:
ok but if I change it to `sizeof(sockaddr_store_t)` so in bringup it goes to 
`tmp->server_addr_len  = s_user_params.addr_len;`

then in `server.cpp` it continues with `socklen_t bind_addr_len = g_fds_array[ifd]->server_addr_len;`
Finally when binding: 
`if (bind(ifd, reinterpret_cast<const sockaddr *>(p_bind_addr), bind_addr_len) < 0)`

if `bind_addr_len` is not set to 108 (`MAX_UDS_NAME`) it won't bind.

agalanin-at-nvidia replied:
Are `sizeof(struct sockaddr_un)` and `sizeof(struct sockaddr_storage_t)` different in our case?

EldarShalev replied:
yes 
sizeof(struct sockaddr_un) = 110 
sizeof(struct sockaddr_storage_t)  = 112 

Maybe because of alignment?

agalanin-at-nvidia replied:
Yes, this is definitely an alignment issue. Try to set `addr_len = sizeof(strruct sockaddr_un)`

agalanin-at-nvidia commented on src/defs.h:
```diff
@@ -477,6 +483,7 @@ struct sockaddr_store_t {
         sa_family_t ss_family;
         sockaddr_in addr4;
         sockaddr_in6 addr6;
+        sockaddr_un addr_un;
```
Do not forget to reduce the size of this structure to store only `MAX_UDS_NAME` bytes of UNIX socket name.

EldarShalev replied:
If it's UNIX family isn't it store only MAX_UDS_NAME?
Not sure how more to optimize it

agalanin-at-nvidia replied:
I agree, `sockaddr_un` contain only 108 characters in path. So now I don't understand why you defining `MAX_UDS_NAME` if we have `UNIX_PATH_MAX` from `linux/un.h` that defined to exactly the same value?

EldarShalev replied:
I can't see the define of `UNIX_PATH_MAX` from `linux/un.h`
![define](https://user-images.githubusercontent.com/44922030/173329716-625f9bce-41eb-40fa-8e1e-7f17027f0437.JPG)

EldarShalev replied:
Here they define it by themselves 

![define2](https://user-images.githubusercontent.com/44922030/173329834-0ea013af-afed-4022-94df-c53d1721ac23.JPG)

agalanin-at-nvidia replied:
I hoped UNIX_PATH_MAX was portable. OK, keep using MAX_UDS_NAME.

agalanin-at-nvidia commented:
bot:retest

agalanin-at-nvidia commented on src/common.h:
```diff
@@ -190,4 +190,25 @@ static inline void sockaddr_set_portn(sockaddr_store_t &addr, uint16_t port)
     }
 }
 
+static inline void copy_relevant_sockaddr_params(sockaddr_store_t &dst_addr, sockaddr_store_t &src_addr)
```
```suggestion
static inline void copy_relevant_sockaddr_params(sockaddr_store_t &dst_addr, const sockaddr_store_t &src_addr)
```

EldarShalev replied:
Done

agalanin-at-nvidia commented on src/common.h:
```diff
@@ -190,4 +190,25 @@ static inline void sockaddr_set_portn(sockaddr_store_t &addr, uint16_t port)
+
+/**
+ * @brief
+ * @return concatenation of unix domaon socket string, useful for multiple binding.
```
```suggestion
 * @return build client socket address for UNIX sockets.
```

agalanin-at-nvidia commented on src/common.h:
```diff
@@ -190,4 +190,25 @@ static inline void sockaddr_set_portn(sockaddr_store_t &addr, uint16_t port)
+ * @brief
+ * @return concatenation of unix domaon socket string, useful for multiple binding.
+ */
+static inline std::string sun_path_to_string(const std::string &sun_path, int pid, int ifd)
+{
+    return sun_path + "_" + std::to_string(pid) + "_" + std::to_string(ifd);
+}
```
```suggestion
static inline std::string unix_sockaddr_to_string(const sockaddr_un *sa)
{
    std::string str(sa->sun_path, sizeof(sa->sun_path));
    size_t len = str.find('\0');
    if (len != std::string::npos)
        str.resize(len);
    return str;
}

static inline std::string build_client_socket_name(const sockaddr_un *sa, int pid, int ifd)
{
    return unix_sockaddr_to_string(sa) + "_" + std::to_string(pid) + "_" + std::to_string(ifd);
}
```

agalanin-at-nvidia commented on src/ip_address.h:
```diff
@@ -79,6 +81,13 @@ class IPAddress {
             break;
+        case AF_UNIX:
+            if (len >= sizeof(sockaddr_un)) {
+                m_addr_un = std::string(reinterpret_cast<const sockaddr_un *>(sa)->sun_path);
```
```suggestion
                m_addr_un = unix_sockaddr_to_string(reinterpret_cast<const sockaddr_un *>(sa));
```

EldarShalev replied:
I agree but again we have the problem with `common.h` -> `defs.h` -> `ip_address.h`

so all the changes for feedfile I can use common.h 
but I can't use `unix_sockaddr_to_string` func in common.h

agalanin-at-nvidia replied:
Try to move constructor to cpp file and include "common.h" inside cpp.

agalanin-at-nvidia commented:
bot:retest

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3114,6 +3103,13 @@ static int set_sockets_from_feedfile(const char *feedfile_name) {
 
+        if (tmp->server_addr.ss_family == AF_UNIX) {
+            s_user_params.tcp_nodelay = false;
+            strcpy(s_user_params.addr.addr_un.sun_path, tmp->server_addr.addr_un.sun_path);
```
```suggestion
            strcnpy(s_user_params.addr.addr_un.sun_path, tmp->server_addr.addr_un.sun_path, MAX_UDS_NAME);
```

agalanin-at-nvidia commented on src/ip_address.h:
```diff
@@ -31,32 +31,34 @@
 #include <sys/socket.h>
 #include <netinet/in.h>
+#include <sys/un.h>      /* definitions for UNIX domain sockets */
+
```
extra empty line

agalanin-at-nvidia reviewed (changes requested):
Good job! We're close to completion. Please re-read all the open comments and fix them (https://github.com/Mellanox/sockperf/pull/178#discussion_r886382033, https://github.com/Mellanox/sockperf/pull/178#discussion_r888662565 etc).

agalanin-at-nvidia commented on src/server.cpp:
```diff
@@ -252,9 +252,10 @@ int Server<IoType, SwitchActivityInfo, SwitchCalcGaps>::server_accept(int ifd) {
+        //fds_data *tmp;
 
-        tmp = (struct fds_data *)MALLOC(sizeof(struct fds_data));
+        // tmp = new fds_data;
```
remove commented code

agalanin-at-nvidia commented on src/server.h:
```diff
@@ -32,6 +32,7 @@
 #include "defs.h"
 #include "common.h"
 #include "input_handlers.h"
+#include <memory>
```
can be moved into server.cpp

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3114,6 +3101,14 @@ static int set_sockets_from_feedfile(const char *feedfile_name) {
 
+        if (tmp->server_addr.ss_family == AF_UNIX) {
+            s_user_params.tcp_nodelay = false;
+            strncpy(s_user_params.addr.addr_un.sun_path, tmp->server_addr.addr_un.sun_path, MAX_UDS_NAME);
+            s_user_params.addr.ss_family = AF_UNIX;
+            tmp->server_addr = s_user_params.addr;
```
You're copying in one direction and then in another. Since we're filling `s_user_params` here, it's enough to do one assignment:
```suggestion
            s_user_params.addr = tmp->server_addr;
```

EldarShalev replied:
yes sorry of course

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3011,7 +2991,12 @@ static int set_sockets_from_feedfile(const char *feedfile_name) {
+            if (regexpres == 0) {
+                type = line[0];
+                addr = std::string(line + m[1].rm_so, line + m[1].rm_eo);
+            }
```
probably you need to have one more closing brace `}` here

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -1725,13 +1725,32 @@ static int resolve_sockaddr(const char *host, const char *port, int sock_type,
+            addr_len = MAX_UDS_NAME;
+            strncpy(tmp->addr_un.sun_path, path.c_str(), MAX_UDS_NAME);
+            tmp->addr_un.sun_family = AF_UNIX;
+            tmp->ss_family = AF_UNIX;
```
no need to set it twice sinсe `ss_family` and `addr_un` is in the same union.

agalanin-at-nvidia commented:
bot:retest

agalanin-at-nvidia commented:
bot:retest

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3035,17 +3017,21 @@ static int set_sockets_from_feedfile(const char *feedfile_name) {
+                        "each line must have the following format: ip:port or type:ip:port or "
+                        "type:ip:port:mc_src_ip or type:/PATH. "
+                        "IPv6 addresses must be enclosed in square brackets."
+                        "Unix domain socket must start with '/'",
```
```suggestion
                        "UNIX domain socket addresses must be an absolute paths (starting with '/').",
```

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3035,17 +3017,21 @@ static int set_sockets_from_feedfile(const char *feedfile_name) {
-            rc = SOCKPERF_ERR_INCORRECT;
-            break;
-        }
+            else {
+                log_msg("Invalid input in line %s: "
+                        "each line must have the following format: ip:port or type:ip:port or "
+                        "type:ip:port:mc_src_ip or type:/PATH. "
+                        "IPv6 addresses must be enclosed in square brackets."
+                        "Unix domain socket must start with '/'",
+                        line);
+                rc = SOCKPERF_ERR_INCORRECT;
+                break;
+            }
```
Please shift this code lock one column left to be aligned with the corresponding `if` blocks.

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3376,13 +3379,13 @@ int bringup(const int *p_daemonize) {
                 if ((curr_fd = (int)socket(tmp->server_addr.ss_family, tmp->sock_type, 0)) <
                     0) { // TODO: use SOCKET all over the way and avoid this cast
-                    log_err("socket(AF_INET4/6, SOCK_x)");
+                    log_err("socket(AF_INET4/6 /AF_UNIX, SOCK_x)");
```
```suggestion
                    log_err("socket(AF_INET4/6/AF_UNIX, SOCK_x)");
```

agalanin-at-nvidia commented on doc/main.dox:
```diff
@@ -103,11 +103,14 @@ sockperf: [tid 4805] using epoll() to block on socket(s)
    - address - Internet Protocol (IP) address or host name (IPv6 addresses must be enclosed in square brackets);
    - port - Port number;
    - mc_src_addr - Optional multicast source IP address or host name.
+   - PATH - path for Unix Domain Socket, the line must start with '/'
```
```suggestion
   - PATH - absolute path for UNIX Domain Socket, the line must start with '/'
```

agalanin-at-nvidia commented on doc/main.dox:
```diff
@@ -103,11 +103,14 @@ sockperf: [tid 4805] using epoll() to block on socket(s)
    or
    [U|T]:address:port:mc_src_addr
+   or
+   [U|T]:/PATH
```
```suggestion
   [U|T]:PATH
```
because / is a part of absolute PATH

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -2948,9 +2919,11 @@ static int set_sockets_from_feedfile(const char *feedfile_name) {
 #else // NEED_REGEX_WORKAROUND
-    const std::regex regexpr(IP_PORT_FORMAT_REG_EXP);
+    const std::regex regexpr_ip(IP_PORT_FORMAT_REG_EXP);
+    const std::regex regexpr_unix(UNIX_DOMAIN_SOCKET_FORMAT_REG_EXP);
```
Please add UDS socket example into a comment before this function (as already done for IPv4/IPv6 addresses.

agalanin-at-nvidia commented:
bot:retest

agalanin-at-nvidia commented on doc/main.dox:
```diff
@@ -116,7 +119,7 @@ sockperf: [tid 4805] using epoll() to block on socket(s)
 @code
  -h,-?   --help,--usage         -Show the help message and exit.
          --tcp                  -Use TCP protocol (default UDP).
```
I think that it will be useful to add a synonym `--stream` for `--tcp` to be consistent with UDS terminology.

EldarShalev replied:
Done

agalanin-at-nvidia replied:
It would be good to change a description too. For example, "Use stream socket/TCP protocol (default is diagram socket/UDP protocol)".
